joern-cli/frontends/javasrc2cpg/src/main/scala/io/joern/javasrc2cpg/typesolvers/noncaching/JdkJarTypeSolver.scala

```diff
@@ -0,0 +1,171 @@
+import org.slf4j.LoggerFactory
+import java.util.jar.JarFile
+import java.io.IOException
+import java.util.jar.JarEntry
+import scala.util.Success
+import scala.util.Try
+import scala.util.Failure
+import io.joern.x2cpg.SourceFiles
+import javassist.CtClass
+import com.github.javaparser.symbolsolver.javassistmodel.JavassistFactory
+import javassist.NotFoundException
+
+class JdkJarTypeSolver private (jdkPath: String) extends TypeSolver {
+
+  private val logger = LoggerFactory.getLogger(this.getClass())
+
+  private var parent: Option[TypeSolver] = None
+  private val classPool                  = new NonCachingClassPool()
+  private val knownClassNames            = mutable.Map[String, String]()
+
+  private type RefType = ResolvedReferenceTypeDeclaration
+
+  override def getParent(): TypeSolver = parent.get
+
+  override def setParent(parent: TypeSolver): Unit = {
+    this.parent match {
+      case None =>
+        this.parent = Some(parent)
+
+      case Some(_) =>
+        throw new RuntimeException("JdkJarTypeSolver parent may only be set once")
+    }
+  }
+
+  override def tryToSolveType(name: String): SymbolReference[ResolvedReferenceTypeDeclaration] = {
+    knownClassNames
+      .get(name)
+      .flatMap(lookupAndConvertClass)
+      .getOrElse(SymbolReference.unsolved(classOf[RefType]))
+  }
+
+  override def solveType(name: String): ResolvedReferenceTypeDeclaration = {
+    tryToSolveType(name) match {
+      case symbolReference if symbolReference.isSolved() =>
+        symbolReference.getCorrespondingDeclaration()
+
+      case _ => throw new UnsolvedSymbolException(name)
+    }
+  }
+
+  private def ctClassToRefType(ctClass: CtClass): RefType = {
+    JavassistFactory.toTypeDeclaration(ctClass, getRoot())
+  }
+
+  private def refTypeToSymbolReference(refType: RefType): SymbolReference[RefType] = {
+    SymbolReference.solved[RefType, RefType](refType)
+  }
+
+  private def lookupAndConvertClass(name: String): Option[SymbolReference[RefType]] = {
+    Try(classPool.get(name)) match {
+      case Success(ctClass) =>
+        val refType      = ctClassToRefType(ctClass)
+        val solvedSymbol = refTypeToSymbolReference(refType)
+        Some(solvedSymbol)
+
+      case Failure(_: NotFoundException) =>
+        logger.error(s"BUG! Could not find class $name in class pool. This is not supposed to be possible!")
+        None
+
+      case Failure(e) =>
+        logger.warn("Unexpected exception getting $name from class pool", e)
+        None
+    }
+  }
+
+  private def addJarToClassPool(path: String): Unit = {
+    path match {
+      case jarPath if jarPath.endsWith(".jar") =>
+        classPool.appendClassPath(jarPath)
+
+      case jmodPath =>
+        classPool.appendClassPath(new JdkArchiveClassPath(jmodPath))
+    }
+  }
+
+  private def withJars(jarPaths: Seq[String]): JdkJarTypeSolver = {
+    jarPaths.foreach { jarPath =>
+      Try(classPool.appendClassPath(new JdkArchiveClassPath(jarPath))) match {
```

Comment:
We don't need to, but I did it this way since the `JdkArchiveClassPath` handles the JDK jars as well (and will any other jars that don't have classes in a `classes.*` package. I changed this for the latest iteration, however, since we could use this class for all jar types in future. The new class path is now `JmodClassPath` to make it clearer.